io-sim-classes/src/Control/Monad/Class/MonadTimer.hs

```diff
@@ -201,20 +207,24 @@ instance MonadTimer IO where
       | otherwise =
         defaultRegisterDelay d
```

Comment:
I accept that this PR fixes the immediate problem, but the underlying problem still exists. 
On a 32 bit machine with a timeout more than ~35min in the future `defaultRegisterDelay` would be invoked which, in turn, makes use of `newTimeout`. This calls `GHC.registerTimeout` where the intrinsic 32 bit constraint rears its ugly head again.

The underlying issue is still there!

Reply:
I didn't aim to fix it here; it is requires a bit more work.